pkg/didcomm/protocol/didexchange/service.go

```diff
@@ -64,18 +96,138 @@ func (s *Service) Handle(msg dispatcher.DIDCommMsg) error {
 	if !current.CanTransitionTo(next) {
 		return fmt.Errorf("invalid state transition: %s -> %s", current.Name(), next.Name())
 	}
-	// TODO: call pre-transition listeners -  Issue: https://github.com/hyperledger/aries-framework-go/issues/140
+
+	// trigger the event if the message type is registered
+	ok, err := s.sendEvent(msg, thid, next)
+	if err != nil {
+		return err
+	}
+
+	// if no events are trigger continue the execution
+	if !ok {
+		return s.handle(&message{Msg: msg, ThreadID: thid, NextStateName: next.Name()})
+	}
+
+	return nil
+}
+
+// startCallbackListener starts the callback listeners. The protocol service will trigger registered events on incoming
+// message types along with a callback channel. These callback channels are received here.
+func (s *Service) startCallbackListener() {
+	go func() {
+		for {
+			select {
+			case msg := <-s.callback[connectionInvite]:
+				s.processCallback(msg)
+			case msg := <-s.callback[connectionRequest]:
+				s.processCallback(msg)
+			case msg := <-s.callback[connectionResponse]:
+				s.processCallback(msg)
+			case msg := <-s.callback[connectionAck]:
+				s.processCallback(msg)
+			}
+		}
+	}()
+}
+
+// processCallback processes the callback events.
+func (s *Service) processCallback(didCommCallback event.DIDCommCallback) {
+	// TODO what to do in case of any failures with callback processing ? - currently, logging the error messages and
```

Comment:
"What do do?"

Probably communicate the error via `DIDCommEvent` (adding an "Err" attribute maybe?)?

Reply:
Or perhaps we should have a separate stream of errors encountered.

Reply:
Created a issue for the TODO https://github.com/hyperledger/aries-framework-go/issues/242